### DISCUSSION TITLE
Use of TimeLib.h with this library demonstrates strange month calculation from TimeLib functions

### DISCUSSION THREAD

**Discussion by Makuna** [Makuna/Rtc#150] [Q&A] [answered]
There is a report that when using the library Time (https://github.com/PaulStoffregen/Time) that some strange interaction can cause that libraries month() to return erroneous value of 13.

given a date and time of `2025/2/28 4:00` that code below would return a month of 13.

```
    RtcDateTime tmp = Rtc.GetDateTime();
    Serial.print("year() - "); Serial.println(year(tmp));
    Serial.print("month()- "); Serial.println(month(tmp));
```

**Comment by Makuna** [comment]
I created a test sketch to ignore the RTC and just work with RtcDateTime and the library in question.  I ran it on ESP32 and didn't see any issues.
```
#include <RtcDateTime.h>
#include <TimeLib.h>

#ifndef countof
#define countof(array) (sizeof(array)/sizeof(array[0]))
#endif

void printDateTime(const RtcDateTime& dt)
{
    char string[40];
 
    snprintf_P(string, 
            countof(string),
            PSTR("RTC %02u/%02u/%04u %02u:%02u:%02u"),
            dt.Month(),
            dt.Day(),
            dt.Year(),
            dt.Hour(),
            dt.Minute(),
            dt.Second() );
    Serial.println(string);

    uint32_t totalSeconds = dt.TotalSeconds();
    uint32_t unixTime = dt.Epoch32Time();

    Serial.println(totalSeconds);
    Serial.println(unixTime);

    snprintf_P(string, 
            countof(string),
            PSTR("UNX %02u/%02u/%04u %02u:%02u:%02u"),
            month(unixTime),
            day(unixTime),
            year(unixTime),
            hour(unixTime),
            minute(unixTime),
            second(unixTime) );
    Serial.println(string);
}

RtcDateTime test(2025, 2, 20, 16, 20, 0) ; // 2025 2/20 4:20pm

void setup() {
    Serial.begin(115200);
    while (!Serial); // wait for serial attach
    
    Serial.println();
    Serial.println("Starting Test...");
    Serial.flush();

    delay(5000);
}

void loop() {
    delay(1000);
    Serial.println();
    printDateTime(test);
    test += 60*60*24; // add a day
}
```
and the output around that date and time 
```
RTC 02/27/2025 16:20:00
793988400
1740673200
UNX 02/27/2025 16:20:00

RTC 02/28/2025 16:20:00
794074800
1740759600
UNX 02/28/2025 16:20:00

RTC 03/01/2025 16:20:00
794161200
1740846000
UNX 03/01/2025 16:20:00
```

**Reply by parallyze** [reply]
Whoa.... couldn't wait and tested this exact sketch using Arduino IDE 1.8.13, Rtc_by_Makuna 2.3.5 and Time 1.6.1 on
Arduino Nano (Atmega328):

```
00:15:04.518 -> Starting Test...
00:15:10.521 -> 
00:15:10.521 -> RTC 02/20/2025 16:20:00
00:15:10.521 -> 793383600
00:15:10.521 -> 1740068400
00:15:10.521 -> UNX 07/12/2025 16:20:00
00:15:11.545 -> 
00:15:11.545 -> RTC 02/20/2025 22:07:44
00:15:11.545 -> 793404464
00:15:11.545 -> 1740089264
00:15:11.545 -> UNX 07/12/2025 22:07:44
00:15:12.522 -> 
00:15:12.522 -> RTC 02/21/2025 03:55:28
00:15:12.522 -> 793425328
00:15:12.522 -> 1740110128
00:15:12.522 -> UNX 07/13/2025 03:55:28
00:15:13.544 -> 
00:15:13.544 -> RTC 02/21/2025 09:43:12
00:15:13.544 -> 793446192
00:15:13.544 -> 1740130992
00:15:13.544 -> UNX 07/13/2025 09:43:12
```
That's nowhere near the 86400 seconds added... more like 20800-ish...  and the UNX date is off.

When including timelib before rtcdatetime this happens:
```
00:23:05.176 -> Starting Test...
00:23:11.157 -> 
00:23:11.157 -> RTC 02/20/2025 16:20:00
00:23:11.157 -> 793383600
00:23:11.157 -> 1740068400
00:23:11.157 -> UNX 02/20/2025 16:20:00
00:23:12.134 -> 
00:23:12.134 -> RTC 02/20/2025 22:07:44
00:23:12.134 -> 793404464
00:23:12.180 -> 1740089264
00:23:12.180 -> UNX 02/20/2025 22:07:44
```
Still the 20800-ish addition each step but UNX date is fixed. 


Now running on a Wemos D1 Mini (ESP8266) using the same Arduino IDE and esp8266 3.1.1 (Boards manager -> esp8266):

```
00:29:53.864 -> Starting Test...
00:29:59.899 -> 
00:29:59.899 -> RTC 02/20/2025 16:20:00
00:29:59.899 -> 793383600
00:29:59.899 -> 1740068400
00:29:59.899 -> UNX 02/20/2025 16:20:00
00:30:00.875 -> 
00:30:00.875 -> RTC 02/21/2025 16:20:00
00:30:00.875 -> 793470000
00:30:00.875 -> 1740154800
00:30:00.875 -> UNX 02/21/2025 16:20:00
00:30:01.900 -> 
00:30:01.900 -> RTC 02/22/2025 16:20:00
00:30:01.900 -> 793556400
00:30:01.900 -> 1740241200
00:30:01.900 -> UNX 02/22/2025 16:20:00
```
Including timelib before/after rtcdatetime doesn't matter in this case, exactly the same (correct) output:

```
00:31:11.649 -> RTC 02/20/2025 16:20:00
00:31:11.649 -> 793383600
00:31:11.649 -> 1740068400
00:31:11.649 -> UNX 02/20/2025 16:20:00
00:31:12.671 -> 
00:31:12.671 -> RTC 02/21/2025 16:20:00
00:31:12.671 -> 793470000
00:31:12.671 -> 1740154800
00:31:12.671 -> UNX 02/21/2025 16:20:00
00:31:13.646 -> 
00:31:13.646 -> RTC 02/22/2025 16:20:00
00:31:13.646 -> 793556400
00:31:13.646 -> 1740241200
00:31:13.646 -> UNX 02/22/2025 16:20:00
```

The IDE I'm using is quite old, so I installed the most recent legacy Arduino IDE (1.8.19) including Time 1.6.1 and
Rtc by Makuna 2.3.5 inside a Win10 virtual machine. This yields different results (Nano):

```
00:41:22.895 -> Starting Test...
00:41:28.910 -> 
00:41:28.910 -> RTC 02/20/2025 16:20:00
00:41:28.910 -> 793383600
00:41:28.910 -> 1740068400
00:41:28.910 -> UNX 02/20/2025 16:20:00
00:41:29.896 -> 
00:41:29.896 -> RTC 02/20/2025 22:07:44
00:41:29.896 -> 793404464
00:41:29.896 -> 1740089264
00:41:29.896 -> UNX 02/20/2025 22:07:44
```

UNX date is shown correct, still the same 20800-ish seconds added each step. Interestingly this is the same with
timelib include before or after rtcdatetime:

```
00:43:38.473 -> Starting Test...
00:43:44.476 -> 
00:43:44.476 -> RTC 02/20/2025 16:20:00
00:43:44.476 -> 793383600
00:43:44.476 -> 1740068400
00:43:44.476 -> UNX 02/20/2025 16:20:00
00:43:45.506 -> 
00:43:45.506 -> RTC 02/20/2025 22:07:44
00:43:45.506 -> 793404464
00:43:45.506 -> 1740089264
00:43:45.506 -> UNX 02/20/2025 22:07:44
00:43:46.520 -> 
00:43:46.520 -> RTC 02/21/2025 03:55:28
00:43:46.520 -> 793425328
00:43:46.520 -> 1740110128
00:43:46.520 -> UNX 02/21/2025 03:55:28
```

So in all cases on Arduino Nano the amount of seconds added was completely off, only the old IDE 1.8.13 did
mess up the UNX date when including timelib after rtcdatetime.

The thing with the order could be caused by the old IDE or by something I messed up over time, can't say
that for sure (been using that installation for quite some time).

Not sure if that's the results you've been waiting for... xD

Do you have a Nano + IDE 1.8.19 to verify (or falsify) what I'm seeing?

Edit:
Uninstalled the IDE 1.8.19 inside the VM, removed  documents/Arduino/libraries and reinstalled 1.8.13:
The UNX date is messed up again when timelib is included after rtcdatetime, so it's not my installation of the IDE
causing this strange thing.

And while at it I uninstalled 1.8.13 and installed 1.8.18 - that one also exhibits the UNX problem. Only using 1.8.19
the importance of the include order is gone.

**Reply by Makuna** [reply]
IDE: v1.8.19
Time: v1.6.1
RTC: 2.3.6 (minor changes post last release but should not affect this)
I think the compiler has some bugs.  If it can't calculate the correct offset by this statement `test += 60*60*24;` then I am worried.  It should be 86500, which is greater than uint16_t; but the += operator only takes a uint32_t; so it sounds like the compiler may not be correctly promoting the type.
try replacing that line with `test += 60*60*24u;` and then try `test += 86500u` 
I may not have a nano, but I will pull out a mega and see what it does.

**Reply by Makuna** [reply]
AVR Board: v1.8.6

Yup, compiler does warn that it is doing the wrong thing when compiler for Mega AVR.
```
...t\RtcDateTime_Unix_Test.ino:59:18: warning: integer overflow in expression [-Woverflow]
     test += 60*60*24; // add a day
```
change to `test += 60ul*60*24; ` forcing the first literal to unsigned long it then correctly calculates the literal; but the time macro is failing still
```
RTC 02/22/2025 04:20:00
793513200
1740198000
UNX 13/03/2025 04:20:00

RTC 02/23/2025 04:20:00
793599600
1740284400
UNX 13/04/2025 04:20:00
```

**Reply by parallyze** [reply]
`test += 60*60*24u` produced the same result as before:
(Nano + IDE 1.8.18, will update to .19 again if requested)
```
02:06:33.349 -> Starting Test...
02:06:39.350 -> 
02:06:39.350 -> RTC 02/20/2025 16:20:00
02:06:39.350 -> 793383600
02:06:39.350 -> 1740068400
02:06:39.350 -> UNX 07/12/2025 16:20:00
02:06:40.336 -> 
02:06:40.336 -> RTC 02/20/2025 22:07:44
02:06:40.336 -> 793404464
02:06:40.336 -> 1740089264
02:06:40.336 -> UNX 07/12/2025 22:07:44
02:06:41.366 -> 
02:06:41.366 -> RTC 02/21/2025 03:55:28
02:06:41.366 -> 793425328
02:06:41.366 -> 1740110128
02:06:41.366 -> UNX 07/13/2025 03:55:28
```

`test += 86400u` adds the correct amount of 24h:
```
02:10:28.312 -> Starting Test...
02:10:34.316 -> 
02:10:34.316 -> RTC 02/20/2025 16:20:00
02:10:34.316 -> 793383600
02:10:34.316 -> 1740068400
02:10:34.316 -> UNX 07/12/2025 16:20:00
02:10:35.300 -> 
02:10:35.300 -> RTC 02/21/2025 16:20:00
02:10:35.300 -> 793470000
02:10:35.300 -> 1740154800
02:10:35.300 -> UNX 07/13/2025 16:20:00
02:10:36.329 -> 
02:10:36.329 -> RTC 02/22/2025 16:20:00
02:10:36.329 -> 793556400
02:10:36.329 -> 1740241200
02:10:36.329 -> UNX 07/14/2025 16:20:00
```
UNX date is off because IDE < 1.8.19 currently and timelib included after rtcdatetime.

`test += 60ul*60*24` results are also correct:
```
02:14:36.753 -> Starting Test...
02:14:42.756 -> 
02:14:42.756 -> RTC 02/20/2025 16:20:00
02:14:42.756 -> 793383600
02:14:42.756 -> 1740068400
02:14:42.756 -> UNX 07/12/2025 16:20:00
02:14:43.740 -> 
02:14:43.740 -> RTC 02/21/2025 16:20:00
02:14:43.740 -> 793470000
02:14:43.740 -> 1740154800
02:14:43.740 -> UNX 07/13/2025 16:20:00
```


By the way, the UNX thing get's pretty funny the longer it's running:
```
02:15:41.870 -> RTC 04/20/2025 16:20:00
02:15:41.870 -> 798481200
02:15:41.870 -> 1745166000
02:15:41.870 -> UNX 13/02/2025 16:20:00
02:15:42.872 -> 
02:15:42.872 -> RTC 04/21/2025 16:20:00
02:15:42.872 -> 798567600
02:15:42.872 -> 1745252400
02:15:42.872 -> UNX 07/72/2025 16:20:00
```
:)))

Edit: No warnings at all (verbose output enabled in Arduino IDE). But I remember that one not being
the most reliable one (PRAGMA/FastLED coming to mind...)

**Reply by Makuna** [reply]
Dug into the `breakTime()` routine to see why its calculating just the month and day of month wrong.
line 193 of Time.cpp `monthLength = monthDays[month];`, the first loop through with month = 0, the returned value is 2!  The monthDays array is corrupted.  
at line 181 I inserted the following to dump the monthDays
```
  for (month = 0; month < 12; month++)
  {
      Serial.print(monthDays[month]);
      Serial.print(" ");
  }
  Serial.println();
  ```
  then right before ` if (time >= monthLength) {` I added the following
  ```
    Serial.print(month);
    Serial.print(" ");
    Serial.print(time);
    Serial.print(" ");
    Serial.println(monthLength);
  ```
  
the output looks like the following every time it loops over for the next day (starting 50 of course is different)...
```
0 0 0 0 0 0 0 0 184 184 184 184 
0 50 0
1 50 28
2 22 0
3 22 0
4 22 0
5 22 0
6 22 0
7 22 0
8 22 184

```
so, the serial dump of monthDays doesn't even match just a few lines down (notice the 28 for the 2nd item in the list even though the serial dump shows 0).
I don't know what to say, the code must be generated incorrectly?

**Reply by Makuna** [reply]
So, I added a local monthDays to the breakTime and changed the code to use it
```
const uint8_t monthDaysLocal[] = { 31,28,31,30,31,30,31,31,30,31,30,31 }; // API starts months from 1, this array starts from 0
```
and this works correctly.  
```
RTC 2/20/2025 4:20:0
793340400
1740025200
UNX 31 28 31 30 31 30 31 31 30 31 30 31 
0 50 31
1 19 28
2/20/2025 4:20:0
```

Now, if that same local monthDays is also static (which it really should be), then it fails again.
```
static const uint8_t monthDaysLocal[] = { 31,28,31,30,31,30,31,31,30,31,30,31 }; // API starts months from 1, this array starts from 0
```

**Reply by parallyze** [reply]
That's quite some strange behavior. This is starting to get outside territory familiar to me, but I think I can follow why that's wierd.

The thing I'm struggling to understand is how this is affected by the Arduino IDE / differences from 1.8.18 (above things happen) and 1.8.19, where this does not happen. According to the git it's a security-release and the merged commits aren't really something that makes me suspicious regarding this error. Both, 1.8.18 and 1.8.19, use gcc7.3.0 as far as I can tell.

You're not testing using the Arduino IDE, are you? 

This whole error doesn't make much sense, maybe that's why I forgotten about it. 

The good thing is: Nothing seems to point into a direction related to your library. Even if the include-order is crucial up to 1.18.18
you've clearly proven something inside timelib is not working as expected.

If I can test anything, please let me know and thanks for looking into this so far!

**Reply by Makuna** [reply, marked as answer]
I am using Arduino IDE.  I have found in the past that flags that are passed to the compiler have changed and that cause unintentional effects and BUGs.  AVR for some reason has these issues popup every few years and I don't know why.  I suspect they are looking for optimizations since it's truly an 8 bit processor with tight memory, and they don't use a wide enough test suit to catch this stuff.

My investigation was to see if I could find a root cause and understand why my library can cause any effect on another as it could popup elsewhere.  But my library is pretty straight forward and the only thing is a PROGMEM array (pretty standard AVR).  The timelib does do some funky things to override standard time header.  I wouldn't use timelib on an avrtiny as its not very memory friendly; but this wasn't an issue in my test.

But at this point, I believe it is a code generation issue that happened to affect timelib and I will let others pursue it more.
